### PR TITLE
 Add node affinity feature enhancements  REP

### PR DESCRIPTION
**Pull request by larrylian**
The value of introducing the Labels mechanism for scheduling was discussed in this REP（[Add labels mechanism] #13 ） before. NodeAffinity is now discussed separately from this REP.

In fact, everyone should know about NodeAffinity before. I think we can reach an agreement on the following points in this REP.

- [ ] Agreed on API format
- [ ] The main implementation plan is agreed
- [ ] Ask specific single-point questions, and then discuss and reach a consensus
- [ ] Some specific details can be discussed during the coding process.

### CODE REVIEW

**Comment by ericl** [comment]
This REP seems like it would be blocked on the autoscaler / GCS refactoring proposal, right @scv119 ? Do we need to have that out first before reviewing this one?

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (ericl) [inline reps/2023-02-03-node-affinity-feature-enhancements.md:280]: What's the index strategy for the more complex expressions such as negation and containment? I wonder if we need to cache the list of valid nodes per task/expression to avoid O(n^2) scheduling slowdowns with many nodes and tasks.

**Comment by larrylian** [comment]
> This REP seems like it would be blocked on the autoscaler / GCS refactoring proposal, right @scv119 ? Do we need to have that out first before reviewing this one?

Thanks for checking it so quickly, Now i'm still just a draft, i finished it today, and i'll @you to help review it.

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (larrylian) [reply reps/2023-02-03-node-affinity-feature-enhancements.md]: This tag indexing strategy is actually very simple, I added it in the REP. The tag index only needs to get the nodes with these key&values.
image

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (stephanie-wang) [inline reps/2023-02-03-node-affinity-feature-enhancements.md:60]: Minor, but seems like we don't really need the `node_affinity` wrapper here?

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (stephanie-wang) [inline reps/2023-02-03-node-affinity-feature-enhancements.md:100]: Instead of introducing a new `node_affinity` wrapper, we can just swap out the `NodeAffinitySchedulingStrategy` implementation to use labels instead.

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (stephanie-wang) [inline reps/2023-02-03-node-affinity-feature-enhancements.md:296]: If there are multiple expressions to match, I think it will be O(kN) (k = num expressions, N = nodes), though, right?

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (stephanie-wang) [inline reps/2023-02-03-node-affinity-feature-enhancements.md:372]: This proposal is more advanced than the autoscaling for NodeAffinity.

NodeAffinity is only used for nodes that are already in the cluster, so I believe the only change there was to make sure that the autoscaler didn't mistakenly start additional nodes for a queued task that wouldn't actually be able to be scheduled there.

Since this proposal is for more flexible labels, we will need the autoscaler to be aware of which nodes it should add or remove from the cluster based on the label constraints of queued tasks. So this actually is something the proposal should cover. Specifically:
- what API changes do we need at the cluster config level and for the interface between the autoscaler and the GCS?
- what will the autoscaler policy look like?
- Handling scalability issues? (If an application uses too many different label policies, we'll have to report these in the task load)

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (stephanie-wang) [inline reps/2023-02-03-node-affinity-feature-enhancements.md:274]: Wouldn't we need to add the label shapes requested by tasks to the ResourceRequests?

**Review by stephanie-wang — COMMENTED** [review]
This looks great, thanks for putting it together!

The main comment we should resolve before approval is about the autoscaling implementation. Ideally we would not block this REP on the autoscaler refactor, but it depends on the complexity of this design. Could you address the comment I left about this, and then we can better determine order of operations?

The other question I had is about semantics for "soft". Can you confirm that soft only considers cluster feasibility of a request's label constraints? I.e., it would still schedule a task to a feasible node even if the node has high load?

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (ericl) [reply reps/2023-02-03-node-affinity-feature-enhancements.md:372]: Yeah I don't think this is quite so simple. The autoscaler runs emulated bin-packing to determine what types of nodes it needs to add to the cluster. For example, suppose a user requests {"dc": "zone-1"}. Then, the autoscaler will need to run its bin packing routine with this request and node labels in mind, in order to determine the nodes to launch: https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/_private/resource_demand_scheduler.py

As you can see, the logic is algorithmically complex.

My assessment is that we probably need to refactor the autoscaler to unify this bin packing routine with the C++ implementation, to avoid having to write label affinity evaluation in both Python and C++.

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (larrylian) [reply reps/2023-02-03-node-affinity-feature-enhancements.md:60]: The benefits of wrapping with node_affinity here are as follows:
1. The label representation in node_affinity can be multiple. So it needs to be wrapped with node_affinity
```
# Multiple label expressions can be filled in at the same time, and the relationship between the expressions is "and". The dispatch must satisfy each expression.
# The actual meaning of this expression is that it must be scheduled to a node with a GPU, and as much as possible to a node with a GPU of the A100 type.
actor_1 = Actor.options(
        scheduling_strategy=node_affinity([
            label_in("gpu_type", ["A100"], true),
            label_exists("gpu_type", false)
        ])
    ).remote()
```
2. After actor_affinity also intends to introduce label expressions. At that time, actor_affinity & node_affinity can reuse the four expressions label_in\label_exist at the same time. The code implementation can be more concise and easier for users to understand.
```
Actor.options(
        scheduling_strategy=node_affinity(label_in(key="node_ip", values=["xxx.xxx.xx.xx"], is_soft=false)).remote()

Actor.options(
        scheduling_strategy=actor_affinity(label_in(key="actor_type", values=["dog"], is_soft=false)).remote()

```

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (larrylian) [reply reps/2023-02-03-node-affinity-feature-enhancements.md:100]: Same as above explanation. Label expressions are planned to be used in more places in the future. For example, ActorAffinity or get a batch of nodes through label. So it's better to wrap it with node_affinity/NodeAffinitySchedulingStrategy here. Then the label indicates that it can be reused in multiple places.
```
# get a batch of nodes through label.
Nodes = GetNodeInfoByLabel([label_in("zone", ["usa", "china"])
```

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (larrylian) [reply reps/2023-02-03-node-affinity-feature-enhancements.md:296]: Yes, In a scene with many expressions, each time an expression is processed, fewer and fewer nodes meet the requirements. Performance is good.

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (larrylian) [reply reps/2023-02-03-node-affinity-feature-enhancements.md:274]: This is for synchronizing the Node Labels information to each raylet node in the Raylet scheduling mode.
1. Putting it in NodeResources can reuse the existing resource synchronization mechanism without creating a new label synchronization mechanism
2. Adding labels here can also allow ActorAffinity to reuse this mechanism.
3. Under the Raylet scheduling mechanism, after the scheduling is completed, the local node needs to confirm whether the final resources meet the requirements. So it is convenient to put it here for final confirmation.

**Comment by larrylian** [comment]
@scv119  @ericl @stephanie-wang 
1. We understand that you also have demands for Runtime Env acceleration, and I have also added a proposal to use Node Labels to achieve Runtime Env acceleration, you can take a look.

2. Do you have any comments on the two REPs Node affinity and Actor Affinity? Can you help advance the progress of these two REPs? This way I can start developing as soon as possible.

 Use Node Labels to achieve Runtime Env acceleration:  
image

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (jjyao) [inline reps/2023-02-03-node-affinity-feature-enhancements.md:343]: Not necessarily saying that I'm proposing this but have we considered label being a k-v pair vs a string. Does a string label have the same expressiveness as a k-v label? A good thing of string label is that it can be easily implemented as custom resources internally.

Comment (larrylian) [reply reps/2023-02-03-node-affinity-feature-enhancements.md:343]: @jjyao 

>have we considered label being a k-v pair vs a string?
We should have discussed this before, and now the form of key-value should refer to the best form of other excellent experience.

1. The (key-value) labels mechanism of K8s is one of the highlights of K8s and has been recognized by countless people.
2. The key-value is simple and can be accepted by users naturally. But at the same time a certain flexibility can be submitted. For example: {gpu_type: "A100"} , {gpu_type: "T100"} . This setting is very natural and intuitive. But if it is replaced by a string: {"gpu_type@A100"},{"gpu_type@T100"}. It will be a little troublesome for users.

> implemented as custom resources internally

This issue has been discussed before. NodeAffinity uses custom_resource internally. Wouldn't be a good solution. It not only destroys the semantics of the original custom_resouce, but also the code implementation will be more complicated (1+1)> 2.

https://github.com/ray-project/enhancements/pull/13/files?short_path=5ae2129#diff-5ae21290dd6f2483719e2557543e9335c1867bea3d1f1cfe516e269d0bd9bc12

image

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (zhe-thoughts) [inline reps/2023-02-03-node-affinity-feature-enhancements.md:25]: This is a required piece of information. I suggest @scv119

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (zhe-thoughts) [inline reps/2023-02-03-node-affinity-feature-enhancements.md:4]: "it can only affinity" -> "it can only support affinity"

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (scv119) [reply reps/2023-02-03-node-affinity-feature-enhancements.md:100]: i think that's option2 right?

Comment (jjyao) [reply reps/2023-02-03-node-affinity-feature-enhancements.md:100]: Not exactly. Option 2 doesn't have `label_in` etc?

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (scv119) [inline reps/2023-02-03-node-affinity-feature-enhancements.md:282]: For node affinity where the node labels are static, we don't need this complicated labels data structure, but only `map<String, String> labels` is sufficient?

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (scv119) [reply reps/2023-02-03-node-affinity-feature-enhancements.md:274]: Since ResourceRequest is also used for autoscaling, seemingly we need to add label requests as well.

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (larrylian) [reply reps/2023-02-03-node-affinity-feature-enhancements.md:100]: @jjyao  Do you mean "scheduling_strategy = NodeAffinitySchedulingStrategy([label_in("zone", ["usa", "china"]])?  

NodeAffinitySchedulingStrategy is replaced by node_affinity mainly for more concise and language style more in line with python.

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (larrylian) [reply reps/2023-02-03-node-affinity-feature-enhancements.md:282]: Yes. map<String, String> labels is sufficient.  Here we can use a concise data structure first. Change it later when we need to expand.

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (larrylian) [reply reps/2023-02-03-node-affinity-feature-enhancements.md:25]: Agree . I will add it.

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (larrylian) [reply reps/2023-02-03-node-affinity-feature-enhancements.md]: done

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (scv119) [reply reps/2023-02-03-node-affinity-feature-enhancements.md:100]: @larrylian we can stick with NodeAffinitySchedulingStrategy for the first version, and improve that in a separate PR.

Comment (larrylian) [reply reps/2023-02-03-node-affinity-feature-enhancements.md:100]: @scv119  Do you mean "scheduling_strategy = NodeAffinitySchedulingStrategy([label_in("zone", ["usa", "china"]])" ?    
It's ok to do that。

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (larrylian) [reply reps/2023-02-03-node-affinity-feature-enhancements.md:274]: Ok I'll try to update a rough design by next week

**Comment by rkooo567** [comment]
oh I think it is the same comment as https://github.com/ray-project/enhancements/pull/22#discussion_r1101633839 actually. So I believe it is addressed

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (jjyao) [inline reps/2023-02-03-node-affinity-feature-enhancements.md:329]:
```suggestion
4. both labels affinity and resources are not satisfied? -> Hang & report schedule failed event and detail unstaisfed reason to exposed to users
```

**Review by jjyao — APPROVED** [review]

File: reps/2023-02-03-node-affinity-feature-enhancements.md
Comment (larrylian) [reply reps/2023-02-03-node-affinity-feature-enhancements.md]: done

**Review by rkooo567 — APPROVED** [review]

**Review by zhe-thoughts — APPROVED** [review]

**Pull request merged**
